### PR TITLE
Ensure that the mrenclave matches the bonding account or the shard to…

### LINKED ISSUE #7: unshield_funds may only be called by registered enclaves

### PR DESCRIPTION
**Pull request by echevrier**
… block malicious enclave

### CODE REVIEW

File: src/tests/tests.rs [hunk 1]
```diff
@@ -476,3 +477,207 @@ fn production_mode_enclave_attest_works_when_sgx_debug_mode_not_allowed() {
...
+            TEST4_CERT.to_vec(),
+            URL.to_vec(),
+        ));
+        assert_eq!(Teerex::enclave_count(), 1);
+
+        assert!(Teerex::shield_funds(
+            Origin::signed(AccountKeyring::Alice.to_account_id()),
+            incognito_account.clone(),
+            100,
+            bonding_account.clone(), //Bonding account (Eoc)
```
Comment (clangenb) [inline src/tests/tests.rs]: What does Eoc mean?

File: src/tests/tests.rs [hunk 2]
```diff
@@ -476,3 +477,207 @@ fn production_mode_enclave_attest_works_when_sgx_debug_mode_not_allowed() {
...
+        assert!(Teerex::shield_funds(
+            Origin::signed(AccountKeyring::Alice.to_account_id()),
+            incognito_account.to_vec(),
+            50,
+            not_bonding_account.clone(), //Bonding account (Eoc)
+        )
+        .is_ok());
+
+        assert!(Teerex::unshield_funds(
+            Origin::signed(signer4.clone()), //Alice incognito account
```
Comment (clangenb) [inline src/tests/tests.rs]: I think this comment is wrong? it is not alice?

File: src/lib.rs [hunk 3]
```diff
@@ -233,6 +237,10 @@ decl_module! {
             ensure!(<EnclaveIndex<T>>::contains_key(&sender),
             "[Teerex]: IPFS state update requested by enclave that is not registered");
 
+            let sender_index = <EnclaveIndex<T>>::get(sender);
+            ensure!(<EnclaveRegistry::<T>>::get(sender_index).mr_enclave.encode() == bonding_account.encode(),
+            "[Teerex]: IPFS state update requested by enclave that doesn't match the bonding accounts");
```
Comment (clangenb) [inline src/lib.rs]: I would like to have new error variants introduced here. String errors miss information when they are returned client side.

Comment (clangenb) [reply src/lib.rs]: Please do that for all string errors that are returned in the pallet.

File: src/tests/tests.rs [hunk 4]
```diff
@@ -476,3 +477,207 @@ fn production_mode_enclave_attest_works_when_sgx_debug_mode_not_allowed() {
...
+        assert_eq!(Teerex::enclave_count(), 0);
+
+        assert!(Teerex::unshield_funds(
+            Origin::signed(signer4.clone()),
+            AccountKeyring::Alice.to_account_id(),
+            51,
+            signer4.clone(),
+            call_hash.clone()
+        )
+        .is_err());
```
Comment (clangenb) [inline src/tests/tests.rs]: I we have return error variants in the pallet, we can also check for the specific error here like we do here:

https://github.com/integritee-network/pallet-teerex/blob/3124348b3be86303f7b83ca159bf9353dcb5306f/src/tests/tests.rs#L185

**Review by clangenb — CHANGES_REQUESTED** [review]
Looks in general good. I like that you added so many tests. However, I do have some comments. :)

File: src/tests/tests.rs (hunk 1 above)
Comment (echevrier) [reply src/tests/tests.rs]: Sorry, I will remove this. It was only for me. to understand the flow. According to the book https://book.integritee.network/token_shielding.html

File: src/tests/tests.rs (hunk 2 above)
Comment (echevrier) [reply src/tests/tests.rs]: Yes. Old comment. I will remove it.

File: src/lib.rs [hunk 5]
```diff
@@ -258,7 +260,19 @@ decl_error! {
...
+        EnclaveIsNotRegistered,
+        ///The bonding account doesn't match the enclave
+        BondingAccountNotMatchEnclave,
+        ///The shard doesn't match the enclave
+        ShardNotMatchEnclave,
+        ///The worker url is too long
+        EnclaveUrlTooLong,
+        ///The RA report is too long
+        RaReportTooLong,
+        ///The enclave doesn't exists
```
Comment (clangenb) [inline src/lib.rs]: Short naming hints: In general, it is advised to use `to be` phrases if possible as opposed to `to do` phrases and leave out the verb. So I have some preferences to `NotMatch`, either:

* `NotEqualTo`
* `NotMatching`

File: src/lib.rs [hunk 6]
```diff
@@ -258,7 +260,19 @@ decl_error! {
...
+        EnclaveDoesntExist,
     }
```
Comment (clangenb) [inline src/lib.rs]: Here also try to remove the `to do` part. In encointer we use `InexistentEnclave`

**Review by clangenb — COMMENTED** [review]
Thanks for the changes, you even fixed some old chores! :)

I do have some naming suggestions. :)

File: src/lib.rs (hunk 5 above)
Comment (echevrier) [reply src/lib.rs]: As only the default shard equals the MRENCLAVE, I prefer NotMatching. See: https://github.com/integritee-network/pallet-teerex/issues/7#issuecomment-937723968
So, what about ShardNotMatchingEnclave and BondingAccountNotMatchingEnclave ?

Comment (brenzi) [reply src/lib.rs]: but the erro should include what isn't matching. `ShardNotMatchingEnclave`. Actually, to be future proof (many shards per mrenclave, we should rather call this error `WrongMrenclaveForShard`

**Review by brenzi — APPROVED** [review]

**Review by clangenb — APPROVED** [review]
Looks good, thank you for also resolving old issues!

**Pull request merged**